src/osdc/Objecter.h

```diff
@@ -1250,6 +1250,58 @@ class Objecter : public md_config_obs_t, public Dispatcher {
   // to be drained by consume_blacklist_events.
   bool blacklist_events_enabled;
   std::set<entity_addr_t> blacklist_events;
+  struct pg_mapping_t {
+    epoch_t epoch = 0;
+    std::vector<int> up;
+    int up_primary = -1;
+    std::vector<int> acting;
+    int acting_primary = -1;
+
+    pg_mapping_t() {}
+    pg_mapping_t(epoch_t epoch, std::vector<int> up, int up_primary,
+                 std::vector<int> acting, int acting_primary)
+               : epoch(epoch), up(up), up_primary(up_primary),
+                 acting(acting), acting_primary(acting_primary) {}
+  };
+  std::shared_mutex pg_mapping_lock;
+  // pool -> pg mapping
+  std::map<int64_t, std::vector<pg_mapping_t>> pg_mappings;
+
+  // convenient accessors
+  bool lookup_pg_mapping(const pg_t& pg, pg_mapping_t* pg_mapping) {
+    std::shared_lock l{pg_mapping_lock};
+    auto it = pg_mappings.find(pg.pool());
+    if (it == pg_mappings.end())
+      return false;
+    auto& mapping_array = it->second;
+    if (pg.ps() >= mapping_array.size())
+      return false;
+    if (mapping_array[pg.ps()].epoch != pg_mapping->epoch) // stale
+      return false;
+    *pg_mapping = mapping_array[pg.ps()];
+    return true;
+  }
+  void update_pg_mapping(unsigned pg_num, const pg_t& pg,
+                         pg_mapping_t&& pg_mapping) {
+    std::lock_guard l{pg_mapping_lock};
+    auto& mapping_array = pg_mappings[pg.pool()];
+    if (mapping_array.size() != pg_num) {
+      // catch both pg_num increasing or decreasing
+      mapping_array.resize(pg_num);
+    }
```

Comment:
i think this is safe since we have an exclusive lock here, but it seems like it fits better under prune_pg_mapping, since the map change is the only time that pg_num will change.

Reply:
yeah, i think it's better to move this so that a bad caller adding in a mapping with a non-existant PG would be better ignored or assert.

Reply:
That was my origin version too.

The reason why I did not move this code block into the `prune_pg_mapping ` helper is that `handle_osd_map ` may need to scan each new map to resend any laggy requests by calling `_calc_target `, too.

To avoid this I have to call `prune_pg_mapping ` before any `_scan_requests `(which seems to be spread everywhere), which makes me very uncomfortable.

In practice we only call resize once per pool (e.g., ignoring pg_num changing), so I wouldn't worry about the cost, and it's more robust, IMHO...

Reply:
That was my origin version too.

The reason why I did not move this code block into the `prune_pg_mapping ` helper is that `handle_osd_map ` may need to scan each new map to resend any laggy requests by calling `_calc_target `, too.

And there are possible callers may directly feed objecter an osdmap in.

To avoid the above cases seems I have to call `prune_pg_mapping` at a couple of different places,
which makes me very uncomfortable.

In practice we only call resize once per pool (e.g., ignoring pg_num changing), so I wouldn't worry about the cost, and it's more robust, IMHO...

Reply:
Yeah, that makes sense!